### PR TITLE
Read a subset of metadata columns

### PR DESCRIPTION
**Pull request by victorlin**
### Description of proposed changes

Reading a subset of metadata columns is straightforward for most subcommands, but requires some big changes in `augur filter`:

- Previously, `--output-metadata` relied on an in-memory pandas DataFrame containing all columns. This PR re-implements that option and similar options using a custom Metadata class to write output metadata based on a streamed representation of the input file.
- Reading a subset of columns for `augur filter` requires knowing which columns are used by `augur filter --query`. Automatic detection is implemented here, but `--query` is generic enough that I'm not confident the detection will always work. In cases where it doesn't, an error is raised.
- A new option `--query-columns` allows the user to specify what columns are used in `--query` along with the expected data types.

See commits for details.

### Related issues

This PR builds on ideas from [discussion on Slack](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1690573498445149) and experimentation in #1242.

### Checklist

- [x] Tests modified for changes in behavior
- [x] Tests added for `--query-columns`
- [x] Checks pass
- [x] Add a message in [CHANGES.md](https://github.com/nextstrain/augur/blob/HEAD/CHANGES.md) summarizing the changes in this PR that are end user focused. Keep headers and formatting consistent with the rest of the file.
- [x] Try running on large datasets ([thread](https://github.com/nextstrain/augur/pull/1294#discussion_r1313613899))

### CODE REVIEW

File: CHANGES.md
Comment (victorlin) [inline CHANGES.md:1]: ## Real-world testing

I ran this against the metadata file produced by ncov-ingest (`s3://nextstrain-data/files/ncov/open/metadata.tsv.zst`) which has 8.5 million rows x 58 columns. I used the following command to sample to 10 random sequences:

```
augur filter \
  --metadata metadata.tsv \
  --subsample-max-sequences 10 \
  --output-metadata out.tsv
```

This took 8m21s to run on `master`, and 6m21s with changes from this PR. Here's [profiling results](https://github.com/nextstrain/augur/files/12501787/profiling.zip) before and after, which I visualized in Snakeviz. A summary:

- Time spent accessing in-memory DataFrames dropped from 494s to 258s.
- Without these changes, `to_csv` takes just a fraction of a second because the metadata for the 10 sequences is already loaded into memory.
- With these changes, it takes 109s to run through the metadata file to find the lines for the 10 sequences that are wanted.

The example command benefits from a net positive improvement in run time. Although writing time increased due to 5173cb7c8ea0f4d161a1f99af6493bad04209d46, reading time decreased even more due to ac23e80376fc517cd50419246db49957f5f43c8a.

This was a "best case scenario" for these changes though, since no metadata columns were used, only strain name. I should probably test with `--group-by`, `--min-date`, and other options that load additional columns to get a better picture.

I did not do any memory profiling. Memory usage is not an issue without these changes, and should be less of an issue with the changes.

File: augur/io/metadata.py
Comment (victorlin) [inline augur/io/metadata.py:27]: This needs a changelog entry.

File: CHANGES.md
Comment (victorlin) [reply CHANGES.md:1]: No, not yet. Still planning to do so before merging.

File: augur/filter/include_exclude_rules.py
Comment (tsibley) [inline augur/filter/include_exclude_rules.py]: My initial expectation was that `--query-columns` would be used to _augment_ the fallback behaviour of "convert all used columns to numeric", not replace it wholesale. Does that make sense? Did you consider that behaviour instead?

I'm also somewhat surprised that we'd raise an error here instead of keeping the ultimate fallback of converting all columns (i.e. since we don't know what is used) to numeric. 

To sum up,  Basically something like:

```python
if column_types is None:
    column_types = {}

variables = extract_variables(query)
if variables is not None:
    columns = variables.intersection(metadata_copy.columns)
else:
    columns = metadata_copy.columns
    # or alternatively: columns = [] to make an exhaustive --query-columns
    # mandatory only when extract_variables() fails

for column in columns:
    column_types.setdefault(column, 'float') # or 'numeric' if taking my other suggestion
```

before the application of `column_types` conversion done above.

File: augur/filter/include_exclude_rules.py
Comment (tsibley) [inline augur/filter/include_exclude_rules.py:197]: I'd expect `int` or `float` to mean actually `int` or `float` and not mean "general numeric conversion". What do you think about allowing `numeric` to mean the `pd.to_numeric(…)` call (i.e. "int or float, whatever it is") and `int` or `float` to mean strict conversion (or maybe truncation) to those specific types? I think that would mean calling `.astype('int')` and `.astype('float')` (or maybe their nullable equivalents, `'Int64'` and `'Float64'`?), possibly with an `errors=…` kwarg too.

Relatedly, do we want to use `errors='ignore'` when the user specifies a specific type for the column? It's friendly in some cases but hostile in others. (Which points to making it specifiable…)

File: augur/filter/io.py
Comment (tsibley) [inline augur/filter/io.py:69]: Ditto here re: `--query-columns` working to augment `extract_variables(query)`.

File: CHANGES.md
Comment (tsibley) [inline CHANGES.md:7]: Seems also worth listing speed ups here (at least generally), if they hold up across the board in your testing. It's useful to know when upgrading might make something faster.

**Review by tsibley — CHANGES_REQUESTED** [review]
Looks pretty good overall! Several small comments and one significant behaviour change to consider.

**Comment by victorlin** [comment]
#1252 changes the same parts of the code and is likely to be merged soon, so I'm going to rebase this PR on top of that.

File: augur/filter/include_exclude_rules.py
Comment (victorlin) [reply augur/filter/include_exclude_rules.py]: I didn't consider this, but it seems better. I'll try it out.

File: augur/io/metadata.py
Comment (victorlin) [reply augur/io/metadata.py:562]: Good catch, I'll see if the condition can be changed to something more explicit than `not row`.

File: augur/filter/include_exclude_rules.py
Comment (victorlin) [reply augur/filter/include_exclude_rules.py:196]: That sounds better. Will update.

File: CHANGES.md
Comment (victorlin) [reply CHANGES.md]: 👍 I'll add a changelog entry after doing more testing.

File: augur/io/metadata.py
Comment (victorlin) [reply augur/io/metadata.py]: Added in 3dd9cfa8984b987c104f6fe25beb36a24370d37d

**Comment by victorlin** [comment]
Force-pushed changes:

- Added new output to tests
- Split 3da1ceda791cb7918b8d29548733aa7ea5193b93 into 5a9fa15875cfad1e676f9a690e83bc595d826af1...d8b5765d0b0302bc3f515bb719eea8b52ce68796
- Optimized usage of `metadata_object` in d8b5765d0b0302bc3f515bb719eea8b52ce68796
- Address metadata column subsetting in other subcommands: d8b5765d0b0302bc3f515bb719eea8b52ce68796...25a704951d667ec3b50777aac69d9af8f8fa4245

Will mark this PR as draft until conversations above are addressed.

File: augur/io/metadata.py
Comment (victorlin) [reply augur/io/metadata.py:562]: It'll be tricky to allow blank lines since the skipping is [done by DictReader](https://github.com/python/cpython/blob/main/Lib/csv.py#L181-L185) and can't be disabled directly ([StackOverflow](https://stackoverflow.com/a/27707768)). Practically, I don't see a use case for keeping them, so I'll just remove this option.

File: augur/io/metadata.py
Comment (victorlin) [reply augur/io/metadata.py:503]: Tests added in [force-push](https://github.com/nextstrain/augur/compare/0038a4fa50c01bdd9128d973fef0f18d24626310..b8842361d5ceaf0c5849b7c00509d0cf083ab42e).

File: augur/filter/include_exclude_rules.py
Comment (victorlin) [reply augur/filter/include_exclude_rules.py]: Updated in [force-push](https://github.com/nextstrain/augur/compare/64a54743a12f744ee873523f4ec40d1986e3ad88..679519ee5092df34f7658b4e677bb8c5b87f69d2). I chose to not make an exhaustive --query-columns mandatory when extract_variables() fails. This keeps the existing behavior of automatic type inference.

https://github.com/nextstrain/augur/blob/679519ee5092df34f7658b4e677bb8c5b87f69d2/augur/filter/include_exclude_rules.py#L196-L211

File: augur/filter/_run.py
Comment (victorlin) [reply augur/filter/_run.py:1]: Updated in d24f0895f7bc8a6a2962e29081d33dc61a2428a1:

> (2) can be prone to issues, though nothing blocking has been brought to our attention.

File: augur/filter/include_exclude_rules.py
Comment (victorlin) [reply augur/filter/include_exclude_rules.py]: Updated in d84c47cc495be54501fdbbdaa62da6539be86834:

https://github.com/nextstrain/augur/blob/679519ee5092df34f7658b4e677bb8c5b87f69d2/augur/filter/include_exclude_rules.py#L213-L228

Comment (victorlin) [reply augur/filter/include_exclude_rules.py]: I chose `errors=raise` since it's user error. Added a test to show what it looks like:

https://github.com/nextstrain/augur/blob/1a2dae9928e72b943dd213552a80bd9392f35b8c/tests/functional/filter/cram/filter-query-columns.t#L47-L56

File: CHANGES.md
Comment (victorlin) [reply CHANGES.md:1]: Tested using the ncov 100k subsample as input to an `augur filter` command I grabbed from an ncov build. Run time was 16s on `master` and 7.37s with these changes ([cProfile files](https://github.com/nextstrain/augur/files/14130245/augur-pr-1294-stats.zip)).

Summary:

- Metadata reading dropped from 4.3s to 2.15s (pandas `readers.py:read`)
- Accessing in-memory DataFrames dropped from 8.63s to 1.127s (pandas `indexing.py`)
- Metadata writing increased from ~0s to 1.35s (`write_metadata_based_outputs`)

```
augur filter \
    --metadata ~/tmp/augur-filter/metadata.tsv.xz \
    --include defaults/include.txt \
    --exclude defaults/exclude.txt \
    --max-date 6M \
    --exclude-where 'region!=Asia' country=China country=India \
    --group-by country year month \
    --subsample-max-sequences 200 \
    --output-strains ~/tmp/augur-filter/strains.txt
```

File: CHANGES.md
Comment (victorlin) [reply CHANGES.md]: Added to changelog:

> * filter, frequencies, refine: Performance improvements to reading the input metadata file. [#1294][] (@victorlin)
>     * For filter, this comes with increased writing times for `--output-metadata` and `--output-strains`. However, net I/O speed still decreased during testing of this change.
> 
> [#1294]: https://github.com/nextstrain/augur/pull/1294

File: CHANGES.md
Comment (victorlin) [reply CHANGES.md:1]: I triggered a ncov GISAID [trial run](https://github.com/nextstrain/ncov/actions/runs/7733059807) which completed successfully in 6h 36m 55s. This is pretty much the same as [another](https://github.com/nextstrain/ncov/actions/runs/7702226205) trial run 2 days before at 6h 40m 27s. I don't know how much variance there is between run times, and I don't want to compare against non-trial runs or older runs (those have additional Slack notifications and different input metadata sizes). So by this comparison alone, there doesn't seem to be a significant performance benefit for the ncov workflow with GISAID configuration.

Comment (tsibley) [reply CHANGES.md:1]: Hmm.

ISTM that last time I looked at ncov's execution profile, by far the slowest step was TreeTime. So not altogether surprising that filter's speed isn't a big impact in the context of a full build.

I grabbed the `benchmarks/subsample_*` files from those runs to get a little more granular insight into differences in wall clock time and max RSS for each subsample rule invocation. 

`avg(after - before)` for wall clock time was -112s, so it shaved roughly 2 min off each subsample step on average. Equivalent for max RSS is -276 (MB, I believe).

**Review by tsibley — APPROVED** [review]
This LGTM. I haven't gone over the new stuff with as fine a toothed comb as my first pass, but I did skim thru it.

**Pull request merged**
